test/Microsoft.AspNet.Mvc.FunctionalTests/ModelBindingFromFormTest.cs

```diff
@@ -0,0 +1,170 @@
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.TestHost;
+using ModelBindingWebSite;
+using ModelBindingWebSite.Controllers;
+using Newtonsoft.Json;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.FunctionalTests
+{
+    public class ModelBindingFromFormTest
+    {
+        private readonly IServiceProvider _services = TestHelper.CreateServices(nameof(ModelBindingWebSite));
+        private readonly Action<IApplicationBuilder> _app = new ModelBindingWebSite.Startup().Configure;
+
+        [Fact]
+        public async Task FromFormAttribute_CustomModelPrefix_ForParameter()
+        {
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+
+            // [FromForm(Name = "customPrefix")] is used to apply a prefix
+            var url = "http://localhost/FromFormAttribute_Company/CreateCompany";
+            var request = new HttpRequestMessage(HttpMethod.Post, url);
+            var nameValueCollection = new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string,string>("customPrefix.Employees[0].Name", "somename"),
+            };
+
+            request.Content = new FormUrlEncodedContent(nameValueCollection);
+
+            // Act
+            var response = await client.SendAsync(request);
+
+            // Assert
+            var body = await response.Content.ReadAsStringAsync();
+            var company = JsonConvert.DeserializeObject<Company>(body);
+
+            var employee = Assert.Single(company.Employees);
+
+            Assert.Equal("somename", employee.Name);
+        }
+
+        [Fact]
+        public async Task FromForm_CustomModelPrefix_ForCollectionParameter()
+        {
+            // Arrange
+            var server = TestServer.Create(_services, _app);
+            var client = server.CreateClient();
+
+            // [FromForm(Name = "customPrefix")] is used to apply a prefix
+            var url =
+               "http://localhost/FromFormAttribute_Company/CreateCompanyFromEmployees";
+            var request = new HttpRequestMessage(HttpMethod.Post, url);
+            var nameValueCollection = new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string, string>("customPrefix[0].EmployeeSSN", "123132131"),
+            };
+            request.Content = new FormUrlEncodedContent(nameValueCollection);
+
+            // Act
+            var response = await client.SendAsync(request);
+
+            // Assert
+            var body = await response.Content.ReadAsStringAsync();
+            var company = JsonConvert.DeserializeObject<Company>(body);
+
+            var employee = Assert.Single(company.Employees);
+            Assert.Equal("123132131", employee.SSN);
+        }
+
+        [Fact]
+        public async Task FromFormAttribute_CustomModelPrefix_ForProperty()
```

Comment:
nit: very similar scenario to test above.  difference is definitely _not_ " collection parameter versus property. should probably have identical name to above test but with either `_InTopLevelCollection` at end of above test name or appending `_InSecondLevelCollection` here.